### PR TITLE
[CALCITE-6586] Some Rules not firing due to RelMdPredicates returning null in VolcanoPlanner

### PR DESCRIPTION
**Pull request by suibianwanwank**
The fundamental problem this PR solves is to make RelMetadataQuery#getPulledUpPredicates(RelSubset) not always return EMPTY, allowing some rules to be applied in VolcanoPlanner.
However, after some rules were applied, there were some other issues that prevented the test from passing completely. So I split it into three commits for easy review

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/interpreter/Bindables.java
Comment (suibianwanwank) [inline core/src/main/java/org/apache/calcite/interpreter/Bindables.java:272]: The push-down fliter is not in the row count calculation, resulting in inaccurate costs

File: innodb/src/main/java/org/apache/calcite/adapter/innodb/InnodbRules.java
Comment (suibianwanwank) [inline innodb/src/main/java/org/apache/calcite/adapter/innodb/InnodbRules.java:189]: In InnodbFilter, condition contains both indexed and non-indexed conditions, but the actual execution only executes the indexed condition, which causes some rules to incorrectly optimize

**Comment by suibianwanwank** [comment]
Some Druid tests have failed, and I will work on fixing them. It may take some time to run these tests locally, as I haven't done that before.

**Comment by mihaibudiu** [comment]
For druid see git@github.com:zabetak/calcite-druid-dataset.git

File: core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java:163]: this plan seems worse, is this expected?

**Review by mihaibudiu — APPROVED** [review]
I can't really tell whether the plans correspond to your expectations after these changes.
Otherwise the changes look fine.

File: core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java
Comment (suibianwanwank) [reply core/src/test/java/org/apache/calcite/test/JdbcAdapterTest.java:163]: Before Plan:
```
JdbcToEnumerableConverter
  JdbcUnion(all=[true])
    JdbcFilter(condition=[=($0, 1)])
      JdbcTableScan(table=[[foodmart, sales_fact_1997]])
    JdbcFilter(condition=[=($0, 1)])
      JdbcTableScan(table=[[foodmart, sales_fact_1998]])
```
After Plan:
```
JdbcToEnumerableConverter
  JdbcProject(product_id=[1], time_id=[$0], customer_id=[$1], promotion_id=[$2], store_id=[$3], store_sales=[$4], store_cost=[$5], unit_sales=[$6])
    JdbcUnion(all=[true])
      JdbcProject(time_id=[$1], customer_id=[$2], promotion_id=[$3], store_id=[$4], store_sales=[$5], store_cost=[$6], unit_sales=[$7])
        JdbcFilter(condition=[=($0, 1)])
          JdbcTableScan(table=[[foodmart, sales_fact_1997]])
      JdbcProject(time_id=[$1], customer_id=[$2], promotion_id=[$3], store_id=[$4], store_sales=[$5], store_cost=[$6], unit_sales=[$7])
        JdbcFilter(condition=[=($0, 1)])
          JdbcTableScan(table=[[foodmart, sales_fact_1998]])
```
In terms of SQL length the plan did get worse. This is because UnionPullUpConstantRule optimizes a RexInputRef to be a constant(Because of the predicate _where product_id = 1_), so what was originally 8 fields for Union's input is now only 7.

**Comment by mihaibudiu** [comment]
can you please fix the conflicts and squash the commits?

**Comment by suibianwanwank** [comment]
> can you please fix the conflicts and squash the commits?

I will resolve these conflicts today. Also, do you think it would be easier to understand if we keep them as separate commit? It might make it clearer.

**Comment by mihaibudiu** [comment]
Having separate commits if they represent independent logical changes to the code.
I think a good rule which tells you whether something could be broken into separate commits is: when you bisect the commits, the code should build and run fine between the commits. Otherwise they should not be separate.

**Comment by mihaibudiu** [comment]
Are you trying to close 4 issues with this PR?
Ideally it should have been 4 separate PRs, especially if they are self-contained.

**Comment by mihaibudiu** [comment]
If no one objects, I will merge this as 4 commits that close 4 issues, but it is easy to miss the fact that the commits address separate issues - it's not the typical workflow. Keeping JIRA and github in sync is a non-trivial amount of work for maintainers, so it's good if we can make the task simpler.

**Comment by suibianwanwank** [comment]
> If no one objects, I will merge this as 4 commits that close 4 issues, but it is easy to miss the fact that the commits address separate issues - it's not the typical workflow. Keeping JIRA and github in sync is a non-trivial amount of work for maintainers, so it's good if we can make the task simpler.

I apologize for the inconvenience caused. I will follow the correct workflow afterward. And I have compiled and tested each commit individually. 
Also, [CALCITE-6611] is only a temporary solution, so it shouldn't be closed yet.

**Comment by mihaibudiu** [comment]
So what should we do about the corresponding commit for 6611? Someone who reads the commit log may infer that the commits solves the problem. Maybe you can remove the issue number from the commit?

**Comment by mihaibudiu** [comment]
To make sure: this PR is closing the following 3 issues: 6615, 6614, and 6586.
Is that correct?

**Comment by suibianwanwank** [comment]
> To make sure: this PR is closing the following 3 issues: 6615, 6614, and 6586. Is that correct?

Yes, that's correct! Also, I tried to resolve the conflict, but this caused some problems regarding pullUpPredicate being slow with too many values. I'll try to look into this.

**Comment by mihaibudiu** [comment]
you want to do this in the current PR? maybe I should just merge it

**Comment by suibianwanwank** [comment]
> you want to do this in the current PR? maybe I should just merge it

Thank you for your interest in this PR. But I don't think it could MERGE yet, as I don't have a good way of passing this test. I'll create a JIRA ISSUE to report this issue and then we'll evaluate what to do. WDYT

**Comment by mihaibudiu** [comment]
certainly, CI has to pass first

**Comment by mihaibudiu** [comment]
I have converted this to draft, please restore it when it's ready for review

File: core/src/main/java/org/apache/calcite/util/Bug.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/util/Bug.java:223]: The title of the commit that introduces this change is misleading: you are using the convention used for a commit which *solves* an issue, but this commit only works around the issue. So I think before merging the title of the commit should be changed.

File: core/src/test/java/org/apache/calcite/test/ScannableTableTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/ScannableTableTest.java:281]: I personally find it very difficult to evaluate whether these plans are correct by inspection (including the ones in XML files). Or even if they are better in any way.
On the other hand this PR is all about plan quality.
So I will have to trust you that the plans are better than they used to be before.

**Review by mihaibudiu — APPROVED** [review]

File: core/src/main/java/org/apache/calcite/util/Bug.java
Comment (suibianwanwank) [reply core/src/main/java/org/apache/calcite/util/Bug.java:223]: Do you have any suggestions for a commit message, I think it might be good to include this in [CALCITE-6586] commit.

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/util/Bug.java:223]: I have no problem with 6586, just with 6328

Comment (mihaibudiu) [reply core/src/main/java/org/apache/calcite/util/Bug.java:223]: The commit message can be "Introduced bug guard for CALCITE-6328."

Comment (suibianwanwank) [reply core/src/main/java/org/apache/calcite/util/Bug.java:223]: Done.

**Comment by mihaibudiu** [comment]
If no one else has comments I plan to merge this as 4 commits, which close 3 issues.

**Pull request merged**
